File: life-os/apps/web/src/features/sprints/model/sprint.ts
import type { LocalDate } from "@lib/localDateTime";

export type SprintStatus = "PLANNED" | "ACTIVE" | "COMPLETED" | "CANCELLED";

export type TaskStatus = "TO_DO" | "IN_PROGRESS" | "BLOCKED" | "DONE" | "CANCELLED";
export type TaskPriority = "P1" | "P2" | "P3" | "P4";

export interface SprintTask {
  readonly id: string;
  readonly sprintId: string;
  readonly taskId: string;
  readonly title: string;
  /** Null when the current owner-scoped Task projection is unavailable. */
  readonly status: TaskStatus | null;
  readonly storyPoints: number;
  readonly projectName?: string;
  readonly priority?: TaskPriority;
  readonly isCommitted: boolean;
  readonly addedAt?: string;
  readonly removedAt?: string;
  readonly carriedOverToSprintId?: string;
}

export type ScopeChangeType = "TASK_ADDED" | "TASK_REMOVED" | "POINTS_CHANGED" | "CAPACITY_CHANGED";

export interface SprintScopeChangeEvent {
  readonly id: string;
  readonly sprintId: string;
  readonly changeType: ScopeChangeType;
  readonly taskId?: string;
  readonly taskTitle?: string;
  readonly pointsDelta?: number;
  readonly reason?: string;
  readonly timestamp: string;
}

export interface Sprint {
  readonly id: string;
  readonly name: string;
  readonly goal?: string;
  readonly startDate: LocalDate;
  readonly endDate: LocalDate;
  readonly status: SprintStatus;
  readonly targetCapacityPoints: number;
  readonly completedStoryPoints: number;
  readonly totalStoryPoints: number;
  readonly retrospectiveNotes?: string;
  readonly whatWentWell?: string;
  readonly whatCouldBeImproved?: string;
  readonly actionItems?: readonly string[];
  readonly completedAt?: string;
  readonly committedTaskCount?: number;
  readonly completedTaskCount?: number;
  readonly addedTaskCount?: number;
  readonly removedTaskCount?: number;
  readonly carriedOverTaskCount?: number;
  readonly createdAt?: string;
  readonly updatedAt?: string;
  readonly version?: number;
}

File: life-os/apps/api/src/main/java/tech/buildwithpartha/lifeos/sprint/domain/Sprint.java
package tech.buildwithpartha.lifeos.sprint.domain;

import java.time.Instant;
import java.time.LocalDate;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.UUID;

public record Sprint(
    UUID id,
    UUID userId,
    String name,
    Optional<String> goal,
    LocalDate startDate,
    LocalDate endDate,
    SprintStatus status,
    int targetCapacityPoints,
    Optional<String> retrospectiveNotes,
    Optional<String> whatWentWell,
    Optional<String> whatCouldBeImproved,
    List<String> actionItems,
    int committedTaskCount,
    int completedTaskCount,
    int addedTaskCount,
    int removedTaskCount,
    int carriedOverTaskCount,
    int totalStoryPoints,
    int completedStoryPoints,
    Optional<Instant> completedAt,
    Instant createdAt,
    Instant updatedAt,
    List<SprintTask> tasks,
    List<SprintEvent> events,
    long version) {
  public Sprint {
    Objects.requireNonNull(id);
    Objects.requireNonNull(userId);
    Objects.requireNonNull(name);
    Objects.requireNonNull(goal);
    Objects.requireNonNull(startDate);
    Objects.requireNonNull(endDate);
    Objects.requireNonNull(status);
    Objects.requireNonNull(retrospectiveNotes);
    Objects.requireNonNull(whatWentWell);
    Objects.requireNonNull(whatCouldBeImproved);
    Objects.requireNonNull(actionItems);
    Objects.requireNonNull(completedAt);
    Objects.requireNonNull(createdAt);
    Objects.requireNonNull(updatedAt);
    Objects.requireNonNull(tasks);
    Objects.requireNonNull(events);
    if (name.isBlank()) {
      throw new IllegalArgumentException("Sprint name must not be blank");
    }
    if (endDate.isBefore(startDate)) {
      throw new IllegalArgumentException("Invalid Sprint dates");
    }
    if (targetCapacityPoints < 0) {
      throw new IllegalArgumentException("Invalid Sprint capacity");
    }
    actionItems = List.copyOf(actionItems);
    tasks = List.copyOf(tasks);
    events = List.copyOf(events);
  }

  public List<SprintTask> activeTasks() {
    return tasks.stream().filter(SprintTask::active).toList();
  }
}
